theano/tensor/nnet/neighbours.py

```diff
@@ -111,6 +111,43 @@ def grad(self, inp, grads):
                 return [neibs2images(gz, neib_shape, x.shape, mode=self.mode),
                         grad_undefined(self, 1, neib_shape),
                         grad_undefined(self, 2, neib_step)]
+
+        if self.mode in ['valid']:
```

Comment:
This case is already implemented. Why do you want to reimplement it? Using scan for this would be very slow.